systemd/systemd.go

```diff
@@ -931,57 +937,121 @@ func (s *systemd) IsActive(serviceName string) (bool, error) {
 	return false, err
 }
 
-func (s *systemd) Stop(serviceNames []string, timeout time.Duration) error {
+func (s *systemd) Stop(serviceNames []string) error {
 	if s.mode == GlobalUserMode {
 		panic("cannot call stop with GlobalUserMode")
 	}
-	if _, err := s.systemctl(append([]string{"stop"}, serviceNames...)...); err != nil {
```

Comment:
Probably this code was originally meant to be used with the `--no-block` option to systemctl.

I sort of am the opinion that we could have just added that parameter here and avoided this rewrite, though that would not have solved the problem of passing the wrong timeouts around everywhere and would not have gotten the end resulting behavior of actually timing out with the timeouts that are configured for the snaps rather than the random ones strewn about our codebase. 

It would still be nice to be able to support a different timeout for systemd operations for things like when snapd needs to restart system services, because what will happen now (and also what was happening before on the branch to be clear) is that we are subject to whatever the default timeout is for that service, but sometimes we may want tighter timeouts for specific situations

Reply:
@anonymouse64 Thank you for the review. 

Just for interest sake, I did start with --no-block. The fundamental issue I encountered was that with --no-block you have no real direct way to know when the systemd stop attempt completes or gives up, which means you need to implement parallel logic to guess that. (I mean, declaring something "Timed out" after a snapd concept of time, while systemd could still be following its complex escalation process, and will still kill it at the end). The other issue is that the timeout specified in the unit file is not simply the effective timeout the systemd escalation process can take. It gets complicated, for example, if you have multiple ExecStop commands. My conclusion is that the only reliable way to give up and report a Timeout, is to have systemd give up. In order to know when that happens,  we need to know when the child process exits.

Reply:
Thanks for the explanation, that makes sense, it would be great if you could leave a comment in the code (can be in a followup too) explaining these caveats and gotchas